eth/gasestimator/gasestimator.go

```diff
@@ -52,15 +53,25 @@ type Options struct {
 // Estimate returns the lowest possible gas limit that allows the transaction to
 // run successfully with the provided context options. It returns an error if the
 // transaction would always revert, or if there are unexpected failures.
-func Estimate(ctx context.Context, call *core.Message, opts *Options, gasCap uint64, exchangeRates common.ExchangeRates, feeCurrencyBalance *big.Int) (uint64, []byte, error) {
+func Estimate(ctx context.Context, call *core.Message, opts *Options, gasCap uint64) (uint64, []byte, error) {
+	// Celo specific: get balance of fee currency if fee currency is specified
+	feeCurrencyBalance := new(big.Int)
+	if call.FeeCurrency != nil {
+		feeCurrencyBalance = getFeeBalance(call, opts)
+	}
+
+	feeCurrencyContext := core.GetFeeCurrencyContext(opts.Header, opts.Config, opts.State)
```

Comment:
Mh, here we re-create the whole context again, even though we already created it in the caller: https://github.com/celo-org/op-geth/pull/356/files#diff-c426ecd2f7d247753b9ea8c1cc003f21fa412661c1f967d203d4edf8163da344L958

Probably passing it around is better for performance.

Reply:
Yes, it was an actual decision.
And:

> Short-circuit for gasEstimation of txs that transfer Celo, with the gas paid with a different feeCurrency
Avoid unnecessary contract calls for rates (every run was calculating it)

was probably a bad chose of words, haha.

The `run` function was calculating that every time, so if the bisect took 12 runs, we would have created that 12 times.

Now, that `run` function, is not creating the context, but the gasestimator.Estimate is.
So yes that's a new one, but I thought that with that change, the gasestimator was going to be self-contained, instead of sending a new parameter. It's reducing the diffs for the estimator calls, and all the celo related things are self contained in the gasestimator, and not in both places

Reply:
If you want I can change it. It's a performance vs code legibility, I don't have a strong opinion on this one. I thought that with the amount saved by the `run` function, one more just for legibility was not going to be a problem